### DISCUSSION TITLE
Forwarding short url parameters to destination when redirecting

### DISCUSSION THREAD

**Discussion by julienarcin** [ash-jc-allen/short-url#93] [Ideas]
Hello,

I would like to be able to forward and merge the url parameters when redirecting.

I mean, I want to create urls of this type:

https://website.com/short/abc?a=b     =>      https://destination.com/long-url-with-parameters?utm_source=test&a=b

It is currently not working. The parameters (a=b) are deleted when redirecting.

I can implement this functionnality and make a PR.

But first, I would like to know your opinions on which option is the best: 

- Should it be the default behavior ? (forwading parameters)
- Should it be customisable in the config.php file globally ? (forward parameters => true/false)
- Should it be implemented on a per-url basis, in database ?

Thanks !

**Comment by ash-jc-allen** [comment]
Hey @julienarcin, this is quite an interesting one 🤔 

If possible, could you give me an example use case of when you might want to do this?

At their core, the short URLs are just returning a redirect (as you already know). Typically, I wouldn't expect redirects to forward query parameters onto the URL that you are being redirected to. But, maybe there's a use case for them here?

I'm also wondering if adding the parameter forwarding defeats the object of having short URLs? They're intended to be short URLs that would be simple for a human to input themselves. I think that by adding the query params to them, they take away the simplicity of them being user-friendly, short URLs and just end up being typical URLs.

Again though, I'm more than open to suggestions on this depending on the different ways people might be using the package 😄

**Comment by julienarcin** [comment]
Hey @ash-jc-allen 

Thanks for this great package, and for the quick answer :)

My use case for this package is to generate tracking urls on my own website with utm parameters.

An example would be:

https://website.com/s/3x3js
=> https://website.com?utm_medium=social&utm_source=facebook&utm_campaign=launch_11_21&utm_content=post_profile_213

The interesting thing is that I have a dynamic homepage which uses the parameters in urls to customise the landing experience for each customer, depending on the information I have on them.

My urls can become even longer, for example:

https://website.com?utm_medium=email&utm_campaign=...&type=web_agency&location=san_francisco

The last 2 parameters can be auto-replaced by my email provider, and will be specific for each customer.

So my use case would be:
- Create a short url to compact the long utm parameters
- Add the dynamic per-customer parameters behind it (eg: https://website.com/s/3x3js?type=web_agency&location=san_francisco)

And the user should land on the huge URL:
https://website.com?utm_medium=social&utm_source=email&utm_campaign=launch_11_21&utm_content=email_242&type=web_agency&location=san_francisco

I can understand if this a very rare case, and doesn't justify being implemented in the main package.

If it's too specific, I will just use my own forked version.

Let me know what you think :)

**Reply by ash-jc-allen** [reply]
Hey! I'm really glad that you like the package 😄 

I think your use case definitely makes sense, and I've just been having a look online to see how other URL shorteners handle this type of thing. It looks like it's pretty common to forward the query parameters on like you want.

So, I'm more than happy to review a PR for this!

Off the top of my head, I think I'd prefer it to work like this:

- Add a new boolean column to the `short_urls` table to determine whether to forward the params.
- Add a new config option that defaults to `false`. (I'm choosing to default it to `false` because it will keep the same behaviour that's there at the moment. This'll prevent anyone updating their dependencies from breaking if they are using something similar to this).
- Add a new method for building short URLs to manually enable/disable the query param forwarding on a case-by-case basis.
- Update the controller to add these parameters if it's enabled for the short URL.

If you think something like that'll work, feel free to have a crack at a PR for it. But I'm open to suggestions though, so if you have a better idea for it, let me know :)

**Reply by ash-jc-allen** [reply]
By the way, if you do add this, please can you target it at the `v6.0.0` branch? I'll bundle this feature in at the same time as adding PHP 8.1 support and dropping PHP 7.* support.

**Reply by julienarcin** [reply]
Ok with the behavior and the PR 👍

I will work on it .

**Reply by julienarcin** [reply]
Hi @ash-jc-allen 

I pushed the pull request #94 

What do you think ?

**Reply by julienarcin** [reply]
As you can see, I have handled the rebuilding of query parameters as a private method in the controller.

This is not as straightforward as we thought first, because there are many cases that need to be taken into account. For example, if the destination url already has query param, we should take them by default and not override them.

Maybe this logic should deservers its own class ? 🤔
